runtime/pkg/activity/kafka.go

```diff
@@ -53,12 +63,42 @@ func NewKafkaSink(brokers, topic string, logger *zap.Logger) (*KafkaSink, error)
 		return nil, err
 	}
 
-	return &KafkaSink{
+	sink := KafkaSink{
 		producer: producer,
 		topic:    topic,
 		logger:   logger,
 		logChan:  logChan,
-	}, nil
+		closedCh: make(chan struct{}),
+	}
+
+	go sink.processProducerEvents(producer, logger)
+
+	return &sink, nil
+}
+
+func (s *KafkaSink) processProducerEvents(producer *kafka.Producer, logger *zap.Logger) {
+	for {
+		select {
+		case e := <-producer.Events():
```

Comment:
I wonder if the `producer.Events` channel might be closed when `s.producer.Close()` is called. Just to be safe, might be worth doing:
```
case e, ok := <-producer.Events():
  if !ok {
    return
  }
  ...
```